server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java

```diff
@@ -617,6 +640,10 @@ public File(StreamInput in) throws IOException {
             reused = in.readBoolean();
         }
 
+        public static File fileWithUnknownLength(String name) {
+            return new File(name, UNKNOWN_LENGTH, false);
```

Comment:
I'm puzzled by this: we should know the lengths of all the files in the snapshot before we start doing any meaningful work in the recovery. Can you explain more clearly why we need to account for files of unknown length?

Reply:
It's possible that the cache starts pre-warming before the file details are known (as we load the snapshot data and prewarm the cache before the shard has recovered) , hence the reason to deal with files with unknown length.

Reply:
Yes, this seems wrong, by the time we start pre-warming we definitely know the sizes of all the files, although we may not have exposed them to the recovery state yet. Put differently, it seems problematic that we are updating the index-phase recovery stats before we enter the index phase of the recovery. Can we defer those stats updates somehow?

Reply:
I thought about triggering the pre-warming phase just after the shard has started, that would solve the issue. But I'm not sure if we want to start pre-warming the cache as soon as possible to minimize the initial search latency. An alternative could be to inspect the `IndexInputStats` once the shard has started and update the `RecoveryState` accordingly.  I think the later approach would simplify things and would avoid to have to keep track of files with unknown length.